// ra-ui-materialui/esm/button/BulkUpdateWithUndoButton.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import PropTypes from 'prop-types';
import ActionUpdate from '@material-ui/icons/Update';
import { alpha } from '@material-ui/core/styles/colorManipulator';
import { makeStyles } from '@material-ui/core/styles';
import { useUpdateMany, useRefresh, useNotify, useUnselectAll, CRUD_UPDATE_MANY, useResourceContext, useListContext, } from 'ra-core';
import Button from './Button';
var useStyles = makeStyles(function (theme) { return ({
    updateButton: {
        color: theme.palette.primary.main,
        '&:hover': {
            backgroundColor: alpha(theme.palette.primary.main, 0.12),
            // Reset on mouse devices
            '@media (hover: none)': {
                backgroundColor: 'transparent',
            },
        },
    },
}); }, { name: 'RaBulkUpdateWithUndoButton' });
var BulkUpdateWithUndoButton = function (props) {
    var selectedIds = useListContext(props).selectedIds;
    var classes = useStyles(props);
    var notify = useNotify();
    var unselectAll = useUnselectAll();
    var refresh = useRefresh();
    var resource = useResourceContext(props);
    var basePath = props.basePath, classesOverride = props.classes, data = props.data, icon = props.icon, label = props.label, onClick = props.onClick, _a = props.onSuccess, onSuccess = _a === void 0 ? function () {
        notify('ra.notification.updated', {
            type: 'info',
            messageArgs: { smart_count: selectedIds.length },
            undoable: true,
        });
        unselectAll(resource);
        refresh();
    } : _a, _b = props.onFailure, onFailure = _b === void 0 ? function (error) {
        notify(typeof error === 'string'
            ? error
            : error.message || 'ra.notification.http_error', {
            type: 'warning',
            messageArgs: {
                _: typeof error === 'string'
                    ? error
                    : error && error.message
                        ? error.message
                        : undefined,
            },
        });
        refresh();
    } : _b, rest = __rest(props, ["basePath", "classes", "data", "icon", "label", "onClick", "onSuccess", "onFailure"]);
    var _c = useUpdateMany(resource, selectedIds, data, {
        action: CRUD_UPDATE_MANY,
        onSuccess: onSuccess,
        onFailure: onFailure,
        undoable: true,
    }), updateMany = _c[0], loading = _c[1].loading;
    var handleClick = function (e) {
        updateMany();
        if (typeof onClick === 'function') {
            onClick(e);
        }
    };
    return (React.createElement(Button, __assign({ onClick: handleClick, label: label, className: classes.updateButton, disabled: loading }, sanitizeRestProps(rest)), icon));
};
var sanitizeRestProps = function (_a) {
    var basePath = _a.basePath, classes = _a.classes, filterValues = _a.filterValues, label = _a.label, selectedIds = _a.selectedIds, onSuccess = _a.onSuccess, onFailure = _a.onFailure, rest = __rest(_a, ["basePath", "classes", "filterValues", "label", "selectedIds", "onSuccess", "onFailure"]);
    return rest;
};
BulkUpdateWithUndoButton.propTypes = {
    basePath: PropTypes.string,
    classes: PropTypes.object,
    label: PropTypes.string,
    resource: PropTypes.string,
    selectedIds: PropTypes.arrayOf(PropTypes.any),
    icon: PropTypes.element,
    data: PropTypes.any.isRequired,
};
BulkUpdateWithUndoButton.defaultProps = {
    label: 'ra.action.update',
    icon: React.createElement(ActionUpdate, null),
};
export default BulkUpdateWithUndoButton;
