// ra-ui-materialui/esm/input/ArrayInput/SimpleFormIteratorItem.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import { Children, cloneElement, isValidElement, useMemo, } from 'react';
import { Typography } from '@material-ui/core';
import classNames from 'classnames';
import FormInput from '../../form/FormInput';
import { useSimpleFormIterator } from './useSimpleFormIterator';
import { SimpleFormIteratorItemContext, } from './SimpleFormIteratorItemContext';
export var SimpleFormIteratorItem = React.forwardRef(function (props, ref) {
    var basePath = props.basePath, children = props.children, classes = props.classes, disabled = props.disabled, disableReordering = props.disableReordering, disableRemove = props.disableRemove, getItemLabel = props.getItemLabel, index = props.index, margin = props.margin, member = props.member, record = props.record, removeButton = props.removeButton, reOrderButtons = props.reOrderButtons, resource = props.resource, source = props.source, variant = props.variant;
    var _a = useSimpleFormIterator(), total = _a.total, reOrder = _a.reOrder, remove = _a.remove;
    // Returns a boolean to indicate whether to disable the remove button for certain fields.
    // If disableRemove is a function, then call the function with the current record to
    // determining if the button should be disabled. Otherwise, use a boolean property that
    // enables or disables the button for all of the fields.
    var disableRemoveField = function (record) {
        if (typeof disableRemove === 'boolean') {
            return disableRemove;
        }
        return disableRemove && disableRemove(record);
    };
    // remove field and call the onClick event of the button passed as removeButton prop
    var handleRemoveButtonClick = function (originalOnClickHandler, index) { return function (event) {
        remove(index);
        if (originalOnClickHandler) {
            originalOnClickHandler(event);
        }
    }; };
    var context = useMemo(function () { return ({
        index: index,
        total: total,
        reOrder: function (newIndex) { return reOrder(index, newIndex); },
        remove: function () { return remove(index); },
    }); }, [index, total, reOrder, remove]);
    return (React.createElement(SimpleFormIteratorItemContext.Provider, { value: context },
        React.createElement("li", { className: classes.line, ref: ref },
            React.createElement("div", null,
                React.createElement("div", { className: classes.indexContainer },
                    React.createElement(Typography, { variant: "body1", className: classes.index }, getItemLabel(index)),
                    !disabled &&
                        !disableReordering &&
                        cloneElement(reOrderButtons, {
                            index: index,
                            max: total,
                            reOrder: reOrder,
                            className: classNames('button-reorder', "button-reorder-" + source + "-" + index),
                        }))),
            React.createElement("section", { className: classes.form }, Children.map(children, function (input, index2) {
                if (!isValidElement(input)) {
                    return null;
                }
                var _a = input.props, source = _a.source, inputProps = __rest(_a, ["source"]);
                return (React.createElement(FormInput, { basePath: input.props.basePath || basePath, input: cloneElement(input, __assign({ source: source
                            ? member + "." + source
                            : member, index: source ? undefined : index2, label: typeof input.props.label ===
                            'undefined'
                            ? source
                                ? "resources." + resource + ".fields." + source
                                : undefined
                            : input.props.label, disabled: disabled }, inputProps)), record: record, resource: resource, variant: variant, margin: margin }));
            })),
            !disabled && !disableRemoveField(record) && (React.createElement("span", { className: classes.action }, cloneElement(removeButton, {
                onClick: handleRemoveButtonClick(removeButton.props.onClick, index),
                className: classNames('button-remove', "button-remove-" + source + "-" + index),
            }))))));
});
